### PR TITLE
Utility analysis for private partition selection

### PR DESCRIPTION
**Pull request by dvadym**

### CODE REVIEW

File: utility_analysis_new/combiners.py [hunk 1]
```diff
@@ -114,37 +118,96 @@ def __add__(self, other):
         return PartitionSelectionAccumulator(moments=moments_self +
                                              moments_other)
 
+    def compute_probability_to_keep(self, eps: float, delta: float,
+                                    max_partitions_contributed: int) -> float:
+        """Computes the probability that this partition is kept."""
```
Comment (jspacek) [inline utility_analysis_new/combiners.py:123]: WDYT about adding a few more comments here like to describe that if probabilities aren't passed, there is an approximation?

File: utility_analysis_new/combiners.py [hunk 2]
```diff
@@ -114,37 +118,96 @@ def __add__(self, other):
...
+        pass
+
 
 @dataclass
 class UtilityAnalysisCountAccumulator:
     count: int
     per_partition_contribution_error: int
     expected_cross_partition_error: float
     var_cross_partition_error: float
-    partition_selection_accumulator: Optional[PartitionSelectionAccumulator]
```
Comment (jspacek) [inline utility_analysis_new/combiners.py:124]: 🎉 I like this restructuring, thank you!

File: utility_analysis_new/dp_engine.py [hunk 3]
```diff
@@ -80,8 +92,6 @@ def _check_utility_analysis_params(params: pipeline_dp.AggregateParams,
                 {pipeline_dp.Metrics.COUNT, pipeline_dp.Metrics.SUM}))
         raise NotImplementedError(
             f"unsupported metric in metrics={not_supported_metrics}")
-    if public_partitions is None:
```
Comment (jspacek) [inline utility_analysis_new/dp_engine.py:83]: The test for this can now be removed in dp_engine_test, it will fail on expected error

**Review by jspacek — CHANGES_REQUESTED** [review]
🎉 Looks great, just failing on the dp_engine_test

File: utility_analysis_new/dp_engine.py (hunk 3 above)
Comment (dvadym) [reply utility_analysis_new/dp_engine.py:83]: Thanks, the test is removed

File: utility_analysis_new/combiners.py (hunk 1 above)
Comment (dvadym) [reply utility_analysis_new/combiners.py]: Thanks, it's a good point. Done

**Review by dvadym — COMMENTED** [review]
Thanks, I've addressed comments. PTAL

**Review by jspacek — APPROVED** [review]
LGTM! 🚀

File: utility_analysis_new/combiners.py (hunk 1 above)
Comment (jspacek) [reply utility_analysis_new/combiners.py]: 🚀  Thank you!

File: utility_analysis_new/dp_engine.py (hunk 3 above)
Comment (jspacek) [reply utility_analysis_new/dp_engine.py:83]: 👍  ✂️

**Pull request merged**